### PR TITLE
storage: in reclock operator, heartbeat handle to hold back the remap since

### LINKED ISSUE #14740: storage: reclock operator is not holding since of remap shard at `0`
We don't currently want to compact the remap shard, because we don't yet have all the wiring set up to do that safely. However, because of how the reclock operator and persist interact we can currently get into a situation where we do downgrade the `since`, which leads to problems when the reclock operator tries to read bindings that are not available anymore.

**Once an environment/storaged hits this bug it will be stuck in a crash loop.**

Description of the problematic behaviour:
 - reclock operator creates a ReadHandle for the remap shard
 - reclock operator clones that and turns the clone into a Listen, that will periodically downgrade it's since
 - reclock operator stashes the original ReadHandle in it's state
 - that original read handle is never touched again. crucially it's not heartbeated periodically
 - original read handle's lease is eventually expired, meaning that "listen" handle is now periodically downgrading the since to something we don't want

Example of this happening in production:

from @pH14: some `storaged` processes (ex. env: 01d730c7-5a61-4d7a-9f3c-63ed01f34ebf-0, pod: storage-u25-0) are crash looping over invalid as_ofs

``` 
2022-09-06T15:15:47.799285Z  INFO storaged: about to bind to 0.0.0.0:2100
2022-09-06T15:15:47.799306Z  INFO storaged: serving internal HTTP server on 0.0.0.0:6878
2022-09-06T15:15:47.805073Z  INFO mz_service::grpc: Starting to listen on 0.0.0.0:2100
2022-09-06T15:15:57.515055Z  INFO mz_service::grpc: GrpcServer: remote client connected
2022-09-06T15:15:57.718716Z  INFO send_operation{operation="GetObject" service="s3"}:provide_credentials{provider=default_chain}:send_operation{operation="AssumeRoleWithWebIdentity" service="sts"}: aws_http::auth: provider returned CredentialsNotLoaded, ignoring
2022-09-06T15:15:58.430681Z  INFO tokio_postgres::connection: NOTICE: relation "consensus" already exists, skipping    
thread 'timely:work-0' panicked at 'invalid as_of: as_of(Antichain { elements: [1662070037000] }) < since(Antichain { elements: [1662070037001] })', src/storage/src/source/reclock.rs:290:9
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at ./rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: mz_storage::source::reclock::ReclockOperator::new::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-032b93d6d7c8edfec-1/materialize/tests/src/storage/src/source/reclock.rs:290:9
   3: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/future/mod.rs:91:19
   4: mz_storage::source::source_reader_pipeline::source_reader_operator::{{closure}}::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-032b93d6d7c8edfec-1/materialize/tests/src/storage/src/source/source_reader_pipeline.rs:224:37
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/future/mod.rs:91:19
   6: <async_stream::async_stream::AsyncStream<T,U> as futures_core::stream::Stream>::poll_next
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/async-stream-0.3.3/src/async_stream.rs:53:17
   7: mz_storage::source::source_reader_pipeline::source_reader_operator::{{closure}}::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-032b93d6d7c8edfec-1/materialize/tests/src/storage/src/source/source_reader_pipeline.rs:420:51
   8: mz_storage::source::util::source::{{closure}}::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-032b93d6d7c8edfec-1/materialize/tests/src/storage/src/source/util.rs:103:17
   9: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build::{{closure}}::{{closure}}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/dataflow/operators/generic/builder_rc.rs:123:31
  10: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/dataflow/operators/generic/builder_rc.rs:162:26
  11: <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/dataflow/operators/generic/builder_raw.rs:204:9
  12: timely::progress::subgraph::PerOperatorState<T>::schedule
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:684:30
  13: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:344:26
  14: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:310:17
  15: timely::progress::subgraph::PerOperatorState<T>::schedule
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:684:30
  16: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:344:26
  17: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:310:17
  18: timely::worker::Wrapper::step::{{closure}}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/worker.rs:754:57
  19: core::option::Option<T>::map
             at ./rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/option.rs:929:29
  20: timely::worker::Wrapper::step
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/worker.rs:754:26
  21: timely::worker::Worker<A>::step_or_park
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/worker.rs:396:38
  22: mz_storage::storage_state::Worker<A>::run_client
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-032b93d6d7c8edfec-1/materialize/tests/src/storage/src/storage_state.rs:134:17
  23: mz_storage::storage_state::Worker<A>::run
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-032b93d6d7c8edfec-1/materialize/tests/src/storage/src/storage_state.rs:110:33
  24: mz_storage::protocol::server::serve::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-032b93d6d7c8edfec-1/materialize/tests/src/storage/src/protocol/server.rs:94:9
  25: timely::execute::execute::{{closure}}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/execute.rs:287:22
  26: timely_communication::initialize::initialize_from::{{closure}}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/communication/src/initialize.rs:316:33
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Stream closed EOF for environment-01d730c7-5a61-4d7a-9f3c-63ed01f34ebf-0/storage-u25-0 (default)
```

Internal discussion thread: https://materializeinc.slack.com/archives/C01CFKM1QRF/p1662735266121499

### PR DESCRIPTION
**Pull request by aljoscha**
Before, we never interacted with the handle that is meant to hold back the since, meaning it could be expired. Now, we make sure to heartbeat the handle whenever we interact with the other handle that we have (inside a listen).

Fixes #14740

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by aljoscha** [comment]
@philip-stoev Do you have opinions on other tests I should add?

**Comment by philip-stoev** [comment]
@aljoscha it this the one that only happened 15 minutes after a storaged restart? Give me today to think about this.

**Comment by aljoscha** [comment]
> @aljoscha it this the one that only happened 15 minutes after a storaged restart? Give me today to think about this.

Yes, it's that one.

**Comment by aljoscha** [comment]
@petrosagg & @philip-stoev Any last words? We should definitely merge this before the next release.

File: src/storage/src/source/reclock.rs [hunk 1]
```diff
@@ -1486,4 +1506,67 @@ mod tests {
...
+        let mut source_upper = HashMap::new();
+
+        // We do multiple rounds of minting. This will downgrade the since of
+        // the internal listen. If we didn't make sure to also heartbeat the
+        // internal handle that holds back the overall remap since the checks
+        // below would fail.
+        //
+        // We do two rounds and advance the time by half the lease timeout in
+        // between so that the "listen handle" will not timeout but the internal
+        // handle used for holding back the since will timout.
```
Comment (petrosagg) [inline src/storage/src/source/reclock.rs:1529]:
```suggestion
        // handle used for holding back the since will timeout.
```

**Review by petrosagg — APPROVED** [review]
Thanks for also writing a unit test!

**Comment by philip-stoev** [comment]
@aljoscha sorry about that. We need to rely on the unit test here, I am unable to contribute anything right now. Maybe in the future we can make the 15-minute interval configurable so that it can be set to a very low value for testing purposes?

**Pull request merged**